numpy/_core/src/multiarray/methods.c

```diff
@@ -892,24 +892,24 @@ static PyObject *
 array_wraparray(PyArrayObject *self, PyObject *args)
 {
     PyArrayObject *arr;
-    PyObject *obj;
+    PyObject *context_ignored = NULL;
+    /*
+     * return_scalar should always be passed, if it is not default to how
+     * this function behaved in older NumPy versions.
+     */
+    int return_scalar = 0;
 
-    if (PyTuple_Size(args) < 1) {
-        PyErr_SetString(PyExc_TypeError,
-                        "only accepts 1 argument");
+    if (!PyArg_ParseTuple(args, "O!|OO&:__array_wrap__",
+                &PyArray_Type, &arr, &context_ignored,
+                &PyArray_OptionalBoolConverter, &return_scalar)) {
         return NULL;
     }
-    obj = PyTuple_GET_ITEM(args, 0);
-    if (obj == NULL) {
-        return NULL;
-    }
-    if (!PyArray_Check(obj)) {
-        PyErr_SetString(PyExc_TypeError,
-                        "can only be called with ndarray object");
-        return NULL;
-    }
-    arr = (PyArrayObject *)obj;
 
+    /*
+     * Subclasses must implement `__array_wrap__` with all the arguments.
```

Comment:
I found this a bit hard to parse. How about
```
* If the array does not have our own type, we always wrap, even
* if `return_scalar` is set, so that we preserve the (presumably
* important) subclass information.
* If the type matches, we honor `return_scalar`, passing through
* PyArray_Return to convert any array with ndim=0 to scalar.
```

Though perhaps even better to just move the comment up, replacing the new comment on l.897?